# Ensemble File EDA
## DSE-260 Group2
Reference: 
https://github.com/OAAS-BEAR/WildFire-Project

Description:
https://docs.google.com/document/d/1ok5mWfWamvWsknZyMnh8Rgp-PrR70jhtapnUTIPyZc0/edit#bookmark=id.k3bp8l8dxry4


# Imports

In [1]:
# import bp3d
# import numpy as np
# import matplotlib.pyplot as plt
import json
import pandas as pd
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# Explore the Ensemble JSON File

In [2]:
with open('uniform-pgml-success.bp3d.json') as f:
    s = f.read()
ens_json = json.loads(s)
ens_json.keys()

dict_keys(['type', 'df'])

In [3]:
ens_json['type']

'ensemble'

In [4]:
# load into Pandas
ens_df_json = json.loads(ens_json['df'])
ens_df = pd.DataFrame().from_dict(ens_df_json)
ens_df.head(5)

sim_time                                               fuel  wind_speed  \
0       600  {'xlen': 600, 'ylen': 600, 'density': 0.7, 'he...       1.000   
1       600  {'xlen': 600, 'ylen': 600, 'density': 0.7, 'he...       1.000   
2       600  {'xlen': 600, 'ylen': 600, 'density': 0.7, 'he...       1.000   
3       600  {'xlen': 600, 'ylen': 600, 'density': 0.7, 'he...       1.000   
4       600  {'xlen': 600, 'ylen': 600, 'density': 0.7, 'he...       1.000   

   wind_direction  surface_moisture  \
0             230             0.050   
1             230             0.050   
2             230             0.050   
3             230             0.050   
4             230             0.050   

                                              output  \
0  {'steps_fire': 1, 'steps_wind': 1, 'energy_atm...   
1  {'steps_fire': 1, 'steps_wind': 1, 'energy_atm...   
2  {'steps_fire': 1, 'steps_wind': 1, 'energy_atm...   
3  {'steps_fire': 1, 'steps_wind': 1, 'energy_atm...   
4  {'steps_fire': 1, 'steps_wind': 1, 'energy_atm...   

                         topo  \
0  {'total_startup_iters': 0}   
1  {'total_startup_iters': 0}   
2  {'total_startup_iters': 0}   
3  {'total_startup_iters': 0}   
4  {'total_startup_iters': 0}   

                                            ignition  \
0  {'perc': 100, 'dat': ['       igntype=    4
',...   
1  {'perc': 100, 'dat': ['       igntype=    5
',...   
2  {'perc': 100, 'dat': ['       igntype=    5
',...   
3  {'perc': 100, 'dat': ['       igntype=    5
',...   
4  {'perc': 100, 'dat': ['       igntype=    5
',...   

                                 run_id st_state        st_status  st_done  
0  80aa1808-adc0-42c1-88e5-abfe3cfc5c41  SUCCESS  Task completed!      100  
1  d7a29471-89ef-4372-a39f-16f3f306edde  SUCCESS  Task completed!      100  
2  85a4e12f-3016-4531-a062-fe5e6dca07eb  SUCCESS  Task completed!      100  
3  dd17e766-419e-439f-91e4-876a48d600bf  SUCCESS  Task completed!      100  
4  8f2da49a-9f67-4cce-b207-875c207b0ce9  SUCCESS  Task completed!      100

## EXPLORE INPUTS: fuel, wind_speed, wind_direction, surface_moisture, topo, ignition

In [5]:
ens_df.describe()

sim_time  wind_speed  wind_direction  surface_moisture  st_done
count  1155.000    1155.000        1155.000          1155.000 1155.000
mean    600.000       6.214         280.000             0.100  100.000
std       0.000       4.630          31.636             0.041    0.000
min     600.000       1.000         230.000             0.050  100.000
25%     600.000       2.000         250.000             0.050  100.000
50%     600.000       5.000         280.000             0.100  100.000
75%     600.000      10.000         310.000             0.150  100.000
max     600.000      15.000         330.000             0.150  100.000

* Each run has 600 time steps.

## Explore the 'fuel' column

In [6]:
ens_df['fuel'].apply(pd.Series).head(5)

xlen    ylen  density  height
0 600.000 600.000    0.700   1.000
1 600.000 600.000    0.700   1.000
2 600.000 600.000    0.700   1.000
3 600.000 600.000    0.700   1.000
4 600.000 600.000    0.700   1.000

In [7]:
ens_df.fuel.apply(pd.Series).describe()

xlen     ylen  density   height
count 1155.000 1155.000 1155.000 1155.000
mean   600.000  600.000    0.700    1.000
std      0.000    0.000    0.000    0.000
min    600.000  600.000    0.700    1.000
25%    600.000  600.000    0.700    1.000
50%    600.000  600.000    0.700    1.000
75%    600.000  600.000    0.700    1.000
max    600.000  600.000    0.700    1.000

* Input Fuel density is unform 600x600x1 with 0.7 density

## Explore the 'wind_speed' column

In [8]:
ens_df.wind_speed.unique()

array([ 1. ,  2. ,  3. ,  5. ,  7.5, 10. , 15. ])

In [9]:
ens_df.groupby('wind_speed').count()

sim_time  fuel  wind_direction  surface_moisture  output  topo  \
wind_speed                                                                   
1.000            165   165             165               165     165   165   
2.000            165   165             165               165     165   165   
3.000            165   165             165               165     165   165   
5.000            165   165             165               165     165   165   
7.500            165   165             165               165     165   165   
10.000           165   165             165               165     165   165   
15.000           165   165             165               165     165   165   

            ignition  run_id  st_state  st_status  st_done  
wind_speed                                                  
1.000            165     165       165        165      165  
2.000            165     165       165        165      165  
3.000            165     165       165        165      165  
5.000            165     165       165        165      165  
7.500            165     165       165        165      165  
10.000           165     165       165        165      165  
15.000           165     165       165        165      165

* Wind Speed varies from 1 to 15 across the simulation runs. Each wind speed has 165 simulation runs.

## Explore the 'wind_direction' column

In [10]:
ens_df.wind_direction.unique()

array([230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330], dtype=int64)

In [11]:
ens_df.groupby(['wind_speed','wind_direction']).count().head(22)

sim_time  fuel  surface_moisture  output  topo  \
wind_speed wind_direction                                                   
1.000      230                   15    15                15      15    15   
           240                   15    15                15      15    15   
           250                   15    15                15      15    15   
           260                   15    15                15      15    15   
           270                   15    15                15      15    15   
           280                   15    15                15      15    15   
           290                   15    15                15      15    15   
           300                   15    15                15      15    15   
           310                   15    15                15      15    15   
           320                   15    15                15      15    15   
           330                   15    15                15      15    15   
2.000      230                   15    15                15      15    15   
           240                   15    15                15      15    15   
           250                   15    15                15      15    15   
           260                   15    15                15      15    15   
           270                   15    15                15      15    15   
           280                   15    15                15      15    15   
           290                   15    15                15      15    15   
           300                   15    15                15      15    15   
           310                   15    15                15      15    15   
           320                   15    15                15      15    15   
           330                   15    15                15      15    15   

                           ignition  run_id  st_state  st_status  st_done  
wind_speed wind_direction                                                  
1.000      230                   15      15        15         15       15  
           240                   15      15        15         15       15  
           250                   15      15        15         15       15  
           260                   15      15        15         15       15  
           270                   15      15        15         15       15  
           280                   15      15        15         15       15  
           290                   15      15        15         15       15  
           300                   15      15        15         15       15  
           310                   15      15        15         15       15  
           320                   15      15        15         15       15  
           330                   15      15        15         15       15  
2.000      230                   15      15        15         15       15  
           240                   15      15        15         15       15  
           250                   15      15        15         15       15  
           260                   15      15        15         15       15  
           270                   15      15        15         15       15  
           280                   15      15        15         15       15  
           290                   15      15        15         15       15  
           300                   15      15        15         15       15  
           310                   15      15        15         15       15  
           320                   15      15        15         15       15  
           330                   15      15        15         15       15

* For each Wind Speed, we have 11 different wind directions (230-330) - 15 runs for each.

## Explore the 'surface_moisture' column

In [12]:
ens_df.surface_moisture.unique()

array([0.05, 0.1 , 0.15])

In [13]:
ens_df.groupby(['wind_speed','wind_direction', 'surface_moisture']).count().head(22)

sim_time  fuel  output  topo  \
wind_speed wind_direction surface_moisture                                 
1.000      230            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           240            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           250            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           260            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           270            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           280            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           290            0.050                    5     5       5     5   
                          0.100                    5     5       5     5   
                          0.150                    5     5       5     5   
           300            0.050                    5     5       5     5   

                                            ignition  run_id  st_state  \
wind_speed wind_direction surface_moisture                               
1.000      230            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           240            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           250            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           260            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           270            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           280            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           290            0.050                    5       5         5   
                          0.100                    5       5         5   
                          0.150                    5       5         5   
           300            0.050                    5       5         5   

                                            st_status  st_done  
wind_speed wind_direction surface_moisture                      
1.000      230            0.050                     5        5  
                          0.100                     5        5  
                          0.150                     5        5  
           240            0.050                     5        5  
                          0.100                     

* For each Wind Speed and Wind Direction ee have 3 different surface moisture values: 0.05 0.1, 0.12 - 5 runs for each.

## Explore the 'topo' column

In [14]:
ens_df['topo'].apply(pd.Series).describe()

total_startup_iters
count             1155.000
mean                 0.000
std                  0.000
min                  0.000
25%                  0.000
50%                  0.000
75%                  0.000
max                  0.000

## Explore the 'ignition' column

In [15]:
ens_df['ignition'].apply(pd.Series)

perc                                                dat
0      100  [       igntype=    4\n,     &aeriallist\n,   ...
1      100  [       igntype=    5\n,        &atvlist\n,   ...
2      100  [       igntype=    5\n,        &atvlist\n,   ...
3      100  [       igntype=    5\n,        &atvlist\n,   ...
4      100  [       igntype=    5\n,        &atvlist\n,   ...
...    ...                                                ...
1150   100  [       igntype=    4\n,     &aeriallist\n,   ...
1151   100  [       igntype=    5\n,        &atvlist\n,   ...
1152   100  [       igntype=    5\n,        &atvlist\n,   ...
1153   100  [       igntype=    5\n,        &atvlist\n,   ...
1154   100  [       igntype=    5\n,        &atvlist\n,   ...

[1155 rows x 2 columns]

In [16]:
ens_df.ignition.apply(pd.Series).describe()

perc
count 1155.000
mean   100.000
std      0.000
min    100.000
25%    100.000
50%    100.000
75%    100.000
max    100.000

In [17]:
ig = ens_df['ignition'].apply(pd.Series)
ig.head(5)

perc                                                dat
0   100  [       igntype=    4\n,     &aeriallist\n,   ...
1   100  [       igntype=    5\n,        &atvlist\n,   ...
2   100  [       igntype=    5\n,        &atvlist\n,   ...
3   100  [       igntype=    5\n,        &atvlist\n,   ...
4   100  [       igntype=    5\n,        &atvlist\n,   ...

In [18]:
len(ig.dat.astype(str).unique())

5

* For each Wind Speed (7) and Wind Direction (11) and Surface Moisture values (3), we have 5 ignition patterns.

## EXLORE THE 'OUTPUT' column

In [19]:
ens_df['output'].apply(pd.Series).head(5)

steps_fire  steps_wind  energy_atmos  fire_energy  fuels_moist
0           1           1          True         True         True
1           1           1          True         True         True
2           1           1          True         True         True
3           1           1          True         True         True
4           1           1          True         True         True

In [20]:
ens_df.output.apply(pd.Series).describe()

steps_fire  steps_wind
count    1155.000    1155.000
mean        1.000       1.000
std         0.000       0.000
min         1.000       1.000
25%         1.000       1.000
50%         1.000       1.000
75%         1.000       1.000
max         1.000       1.000

In [21]:
ens_df.output.apply(pd.Series).energy_atmos.unique()

array([ True])

In [22]:
ens_df.output.apply(pd.Series).fire_energy.unique()

array([ True])

In [23]:
ens_df.output.apply(pd.Series).fuels_moist.unique()

array([ True])

## CONCLUSION:

__Overall all this ensemble has 1155 simulation runs:__ 

*Inputs:*
- Fuel Density (1 unique value)  
- Wind Speed (7 unique values)  
- Wind Direction (11 unique values)  
- Surface Moisture (3 unique values)  
- Topography (1 unique value)  
- Ignition Patterns (5 unique values)  
-------------------------------------  
Total: 1 * 7 * 11 * 3 * 1 * 5 = __1155__ combinations (=simulation runs)  
Each run has 600 time steps.
    
*Output:*
- Output will be uniform (same structure) for all runs.